File: test/ModularPipelines.UnitTests/Dependencies/DependsOnAllInheritingFromTests.cs
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Time.Testing;
using ModularPipelines.Attributes;
using ModularPipelines.Context;
using ModularPipelines.Engine;
using ModularPipelines.Extensions;
using ModularPipelines.Modules;
using ModularPipelines.TestHelpers;

namespace ModularPipelines.UnitTests.Dependencies;

public class DependsOnAllInheritingFromTests : TestBase
{
    private static readonly TimeSpan ModuleDelay = TimeSpan.FromMilliseconds(50);

    private abstract class BaseModule : Module<bool>
    {
        protected internal abstract override Task<bool> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken);
    }

    // Generic base module for testing open generic type dependencies (Issue #1337)
    private abstract class GenericBaseModule<T> : Module<T>;

    private class GenericModule1 : GenericBaseModule<int>
    {
        protected internal override async Task<int> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Delay(ModuleDelay, cancellationToken);
            return 42;
        }
    }

    private class GenericModule2 : GenericBaseModule<string>
    {
        protected internal override async Task<string> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Delay(ModuleDelay, cancellationToken);
            return "test";
        }
    }

    // This module depends on all modules inheriting from the OPEN generic type GenericBaseModule<>
    [DependsOnAllModulesInheritingFrom(typeof(GenericBaseModule<>))]
    private class DependsOnOpenGenericModule : Module<bool>
    {
        protected internal override async Task<bool> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Yield();
            return true;
        }
    }

    private class Module1 : BaseModule
    {
        protected internal override async Task<bool> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Delay(ModuleDelay, cancellationToken);
            return true;
        }
    }

    [ModularPipelines.Attributes.DependsOn<Module1>]
    private class Module2 : BaseModule
    {
        protected internal override async Task<bool> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Delay(ModuleDelay, cancellationToken);
            return true;
        }
    }

    [ModularPipelines.Attributes.DependsOn<Module1>(Optional = true)]
    private class Module3 : BaseModule
    {
        protected internal override async Task<bool> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Delay(ModuleDelay, cancellationToken);
            return true;
        }
    }

    [ModularPipelines.Attributes.DependsOnAllModulesInheritingFrom<BaseModule>]
    private class Module4 : Module<bool>
    {
        protected internal override async Task<bool> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Yield();
            return true;
        }
    }

    [Test]
    public async Task No_Exception_Thrown_When_Dependent_Module_Present()
    {
        var timeProvider = new FakeTimeProvider();

        var host = await TestPipelineHostBuilder.Create(new TestHostSettings(), timeProvider)
            .AddModule<Module1>()
            .AddModule<Module2>()
            .AddModule<Module3>()
            .AddModule<Module4>()
            .BuildHostAsync();

        var pipelineSummary = await host.ExecutePipelineAsync();
        var resultRegistry = host.RootServices.GetRequiredService<IModuleResultRegistry>();

        var result1 = resultRegistry.GetResult(typeof(Module1))!;
        var result2 = resultRegistry.GetResult(typeof(Module2))!;
        var result3 = resultRegistry.GetResult(typeof(Module3))!;
        var result4 = resultRegistry.GetResult(typeof(Module4))!;

        await Assert.That(result4.ModuleStart).IsGreaterThanOrEqualTo(result1.ModuleStart.Add(ModuleDelay.Add(TimeSpan.FromMilliseconds(-25))));
        await Assert.That(result4.ModuleStart).IsGreaterThanOrEqualTo(result1.ModuleEnd);

        await Assert.That(result4.ModuleStart).IsGreaterThanOrEqualTo(result2.ModuleStart.Add(ModuleDelay.Add(TimeSpan.FromMilliseconds(-25))));
        await Assert.That(result4.ModuleStart).IsGreaterThanOrEqualTo(result2.ModuleEnd);

        await Assert.That(result4.ModuleStart).IsGreaterThanOrEqualTo(result3.ModuleStart.Add(ModuleDelay.Add(TimeSpan.FromMilliseconds(-25))));
        await Assert.That(result4.ModuleStart).IsGreaterThanOrEqualTo(result3.ModuleEnd);
    }

    [Test]
    public async Task DependsOnAllModulesInheritingFrom_Works_With_Open_Generic_Types()
    {
        // Regression test for Issue #1337
        // DependsOnAllModulesInheritingFrom(typeof(GenericBaseModule<>)) should wait for
        // GenericModule1 : GenericBaseModule<int> and GenericModule2 : GenericBaseModule<string>
        var timeProvider = new FakeTimeProvider();

        var host = await TestPipelineHostBuilder.Create(new TestHostSettings(), timeProvider)
            .AddModule<GenericModule1>()
            .AddModule<GenericModule2>()
            .AddModule<DependsOnOpenGenericModule>()
            .BuildHostAsync();

        var pipelineSummary = await host.ExecutePipelineAsync();
        var resultRegistry = host.RootServices.GetRequiredService<IModuleResultRegistry>();

        var result1 = resultRegistry.GetResult(typeof(GenericModule1))!;
        var result2 = resultRegistry.GetResult(typeof(GenericModule2))!;
        var dependentResult = resultRegistry.GetResult(typeof(DependsOnOpenGenericModule))!;

        // The dependent module should start AFTER both generic modules complete
        await Assert.That(dependentResult.ModuleStart).IsGreaterThanOrEqualTo(result1.ModuleEnd);
        await Assert.That(dependentResult.ModuleStart).IsGreaterThanOrEqualTo(result2.ModuleEnd);
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Dependencies/DependsOnAllInheritingFromTests.fs
namespace ModularPipelines.UnitTests.FSharp.Dependencies

open System
open System.Threading.Tasks
open Microsoft.Extensions.DependencyInjection
open Microsoft.Extensions.Time.Testing
open ModularPipelines.Attributes
open ModularPipelines.Context
open ModularPipelines.Engine
open ModularPipelines.Extensions
open ModularPipelines.Modules
open ModularPipelines.TestHelpers
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

[<AutoOpen>]
module private DependsOnAllInheritingFromTestsHelpers =
    [<Literal>]
    let toleranceMilliseconds = -25.0

    let moduleDelay = TimeSpan.FromMilliseconds(50.0)


[<AbstractClass>]
type InheritingBaseModule() =
    inherit Module<bool>()

[<AbstractClass>]
type InheritingGenericBaseModule<'T>() =
    inherit Module<'T>()

type InheritingGenericModule1() =
    inherit InheritingGenericBaseModule<int>()

    override _.ExecuteAsync(_, cancellationToken) =
        task {
            do! Task.Delay(moduleDelay, cancellationToken)
            return 42
        }

type InheritingGenericModule2() =
    inherit InheritingGenericBaseModule<string>()

    override _.ExecuteAsync(_, cancellationToken) =
        task {
            do! Task.Delay(moduleDelay, cancellationToken)
            return "test"
        }

[<DependsOnAllModulesInheritingFrom(typedefof<InheritingGenericBaseModule<_>>)>]
type InheritingDependsOnOpenGenericModule() =
    inherit Module<bool>()

    override _.ExecuteAsync(_, _) =
        task {
            do! Task.Yield()
            return true
        }

type InheritingModule1() =
    inherit InheritingBaseModule()

    override _.ExecuteAsync(_, cancellationToken) =
        task {
            do! Task.Delay(moduleDelay, cancellationToken)
            return true
        }

[<DependsOn(typeof<InheritingModule1>)>]
type InheritingModule2() =
    inherit InheritingBaseModule()

    override _.ExecuteAsync(_, cancellationToken) =
        task {
            do! Task.Delay(moduleDelay, cancellationToken)
            return true
        }

[<ModularPipelines.Attributes.DependsOn(typeof<InheritingModule1>, Optional = true)>]
type InheritingModule3() =
    inherit InheritingBaseModule()

    override _.ExecuteAsync(_, cancellationToken) =
        task {
            do! Task.Delay(moduleDelay, cancellationToken)
            return true
        }

[<DependsOnAllModulesInheritingFrom(typeof<InheritingBaseModule>)>]
type InheritingModule4() =
    inherit Module<bool>()

    override _.ExecuteAsync(_, _) =
        task {
            do! Task.Yield()
            return true
        }

type DependsOnAllInheritingFromTests() =
    inherit TestBase()

    [<Test>]
    member _.No_Exception_Thrown_When_Dependent_Module_Present() = async {
        let timeProvider = FakeTimeProvider()

        let! host =
            TestPipelineHostBuilder.Create(TestHostSettings(), timeProvider)
                .AddModule<InheritingModule1>()
                .AddModule<InheritingModule2>()
                .AddModule<InheritingModule3>()
                .AddModule<InheritingModule4>()
                .BuildHostAsync()
            |> Async.AwaitTask

        do! host.ExecutePipelineAsync() |> Async.AwaitTask |> Async.Ignore

        let resultRegistry = host.RootServices.GetRequiredService<IModuleResultRegistry>()
        let result1 = resultRegistry.GetResult(typeof<InheritingModule1>)
        let result2 = resultRegistry.GetResult(typeof<InheritingModule2>)
        let result3 = resultRegistry.GetResult(typeof<InheritingModule3>)
        let result4 = resultRegistry.GetResult(typeof<InheritingModule4>)

        do! check(Assert.That(result4.ModuleStart >= result1.ModuleStart.Add(moduleDelay.Add(TimeSpan.FromMilliseconds(toleranceMilliseconds)))).IsTrue())
        do! check(Assert.That(result4.ModuleStart >= result1.ModuleEnd).IsTrue())
        do! check(Assert.That(result4.ModuleStart >= result2.ModuleStart.Add(moduleDelay.Add(TimeSpan.FromMilliseconds(toleranceMilliseconds)))).IsTrue())
        do! check(Assert.That(result4.ModuleStart >= result2.ModuleEnd).IsTrue())
        do! check(Assert.That(result4.ModuleStart >= result3.ModuleStart.Add(moduleDelay.Add(TimeSpan.FromMilliseconds(toleranceMilliseconds)))).IsTrue())
        do! check(Assert.That(result4.ModuleStart >= result3.ModuleEnd).IsTrue())
    }

    [<Test>]
    member _.DependsOnAllModulesInheritingFrom_Works_With_Open_Generic_Types() = async {
        let timeProvider = FakeTimeProvider()

        let! host =
            TestPipelineHostBuilder.Create(TestHostSettings(), timeProvider)
                .AddModule<InheritingGenericModule1>()
                .AddModule<InheritingGenericModule2>()
                .AddModule<InheritingDependsOnOpenGenericModule>()
                .BuildHostAsync()
            |> Async.AwaitTask

        do! host.ExecutePipelineAsync() |> Async.AwaitTask |> Async.Ignore

        let resultRegistry = host.RootServices.GetRequiredService<IModuleResultRegistry>()
        let result1 = resultRegistry.GetResult(typeof<InheritingGenericModule1>)
        let result2 = resultRegistry.GetResult(typeof<InheritingGenericModule2>)
        let dependentResult = resultRegistry.GetResult(typeof<InheritingDependsOnOpenGenericModule>)

        do! check(Assert.That(dependentResult.ModuleStart >= result1.ModuleEnd).IsTrue())
        do! check(Assert.That(dependentResult.ModuleStart >= result2.ModuleEnd).IsTrue())
    }
